pallets/liquidity-pools-gateway/src/message_processing.rs

```diff
@@ -0,0 +1,498 @@
+#[derive(Clone)]
+pub struct InboundProcessingInfo<T: Config> {
+	pub domain_address: DomainAddress,
+	pub router_ids: Vec<T::RouterId>,
+	pub current_session_id: T::SessionId,
+	pub expected_proof_count_per_message: u32,
+}
+
+impl<T: Config> Pallet<T> {
+	/// Retrieves all available routers for a domain and then filters them based
+	/// on the routers that we have in storage.
+	pub fn get_router_ids_for_domain(domain: Domain) -> Result<Vec<T::RouterId>, DispatchError> {
+		let all_routers_for_domain = T::RouterProvider::routers_for_domain(domain);
+
+		let stored_routers = Routers::<T>::get().ok_or(Error::<T>::RoutersNotFound)?;
+
+		let res = all_routers_for_domain
+			.iter()
+			.filter(|x| stored_routers.iter().any(|y| *x == y))
+			.map(|x| x.clone())
+			.collect::<Vec<_>>();
+
+		if res.is_empty() {
+			return Err(Error::<T>::NotEnoughRoutersForDomain.into());
+		}
+
+		Ok(res)
+	}
+
+	/// Calculates and returns the proof count required for processing one
+	/// inbound message.
+	fn get_expected_proof_count(router_ids: &Vec<T::RouterId>) -> Result<u32, DispatchError> {
+		let expected_proof_count = router_ids
+			.len()
+			.ensure_sub(1)
+			.map_err(|_| Error::<T>::NotEnoughRoutersForDomain)?;
+
+		Ok(expected_proof_count as u32)
+	}
+
+	/// Gets the message proof for a message.
+	fn get_message_proof(message: T::Message) -> Proof {
+		match message.get_message_proof() {
+			None => message
+				.to_message_proof()
+				.get_message_proof()
+				.expect("message proof ensured by 'to_message_proof'"),
+			Some(proof) => proof,
+		}
+	}
+
+	/// Creates an inbound entry based on whether the inbound message is a
+	/// proof or not.
+	fn create_inbound_entry(
+		domain_address: DomainAddress,
+		message: T::Message,
+		expected_proof_count: u32,
+	) -> InboundEntry<T> {
+		match message.get_message_proof() {
+			None => InboundEntry::Message {
+				domain_address,
+				message,
+				expected_proof_count,
+			},
+			Some(_) => InboundEntry::Proof { current_count: 1 },
+		}
+	}
+
+	/// Validation ensures that:
+	///
+	/// - the router that sent the inbound message is a valid router for the
+	///   specific domain.
+	/// - messages are only sent by the first inbound router.
+	/// - proofs are not sent by the first inbound router.
+	fn validate_inbound_entry(
+		inbound_processing_info: &InboundProcessingInfo<T>,
+		router_id: &T::RouterId,
+		inbound_entry: &InboundEntry<T>,
+	) -> DispatchResult {
+		let router_ids = inbound_processing_info.router_ids.clone();
+
+		ensure!(
+			router_ids.iter().any(|x| x == router_id),
+			Error::<T>::UnknownRouter
+		);
+
+		match inbound_entry {
+			InboundEntry::Message { .. } => {
+				ensure!(
+					router_ids.get(0) == Some(&router_id),
+					Error::<T>::MessageExpectedFromFirstRouter
+				);
+
+				Ok(())
+			}
+			InboundEntry::Proof { .. } => {
+				ensure!(
+					router_ids.get(0) != Some(&router_id),
+					Error::<T>::ProofNotExpectedFromFirstRouter
+				);
+
+				Ok(())
+			}
```

Comment:
Not fully get these conditions. Why is there a restriction on receiving the message from the first router in the list and the proofs for the rest? Is it a requirement?

Reply:
I mean, I understand for sending message we should choose one of them, let's say, the first one, but why for the inbound messages we should expect the first router in the list? Could not by any of them?

Reply:
Requirement as per [slack discussion](https://kflabs.slack.com/archives/C06FRHCRQL9/p1723054166639529?thread_ts=1723038229.117829&cid=C06FRHCRQL9).

Reply:
Thanks for the ref. 

I've re-read the thread, and I still do not see why the router configured with index 0 is the router that must send the message. From what I understand from the conversation is that a router can not send a message and also a proof, they must be different routers. But the router who send the message can be router with index 1 and the router who send the proof, router with index 0, isn't?

Reply:
I'm not sure if this is a hard requirement or not. @hieronx can you please clarify?

Reply:
The design is such that 
- The set of routers on the EVM side to Centrifuge, must match the set of routers from Centrifuge that that EVM domain
- The first router in the list must send the message, and the remaining routers must send the proofs

Thus it logically follows that the incoming handler should only accept messages from the first router and proofs from the other routers.

Reply:
So, when configuring routers, the configuration order matters 👍🏻